rfcs/20240228-graph-api-support-accumulation-mode/README.md

```diff
@@ -0,0 +1,258 @@
+
+## Accumulation mode definition
+
+oneDNN graph will reuse the C and C++ API for accumulation mode enumerations defined by primitive API:
+
+C API:
+
+```c
+typedef enum {
+    dnnl_accumulation_mode_strict,  
+    dnnl_accumulation_mode_relaxed, 
+    dnnl_accumulation_mode_any,     
+    dnnl_accumulation_mode_s32,     
+    dnnl_accumulation_mode_f32,     
+    dnnl_accumulation_mode_f16,     
+} dnnl_accumulation_mode_t;
+```
+
+C++ API:
+
+```cpp
+enum class accumulation_mode {
+    strict = dnnl_accumulation_mode_strict,
+    relaxed = dnnl_accumulation_mode_relaxed,
+    any = dnnl_accumulation_mode_any,
+    s32 = dnnl_accumulation_mode_s32,
+    f32 = dnnl_accumulation_mode_f32,
+    f16 = dnnl_accumulation_mode_f16,
+}
+```
+
+The accumulation mode attribute accepts:
+- `strict` (default): For floating-point datatypes, the default accumulation datatype is `f32`
+(or f64 for f64 primitives). For integral datatypes, the default accumulation datatype is `s32`.
+- `relaxed`: Same as strict except some partial accumulators can be
+  rounded to the src/dst datatype in memory.
+- `any`: Uses the fastest implementation available with one of the
+  src/dst datatypes or a higher precision accumulation datatype.
+- `f32`, `f16` and `s32`: Uses the specified accumulation datatype.
+
+ Framework users may need map their own accumulation mode definitions to the above enumerations.
+
+## Proposal
+
+### Option 1: Specify accumulation mode thorugh global function
+
+Use a global function to set the accumulation mode, which will affect all the created graphs. Users should set the accumulation mode at the graph build stage before compilation.
+
+The C API will be like:
+
+```c
+status_t DNNL_GRAPH_API
+dnnl_graph_set_accumulation_mode(dnnl_graph_accumulation_mode_t mode);
+```
+
+The correspodning C++ API will be like:
+
+```cpp
+namespace dnnl {
+namespace graph {
+void set_accumulation_mode(accumulation_mode mode);
+}
+}
+```
+
+#### Pros and Cons
+
+Pros:
+
+- No change on all existing APIs
+
+Cons:
+
+- Only provide coarse-grained control on all graphs
+- Need efforts for managing global status.
+- Require users to provide essential information at the early graph build stage 
+
+### Option 2: Support accumulation mode on graph level( Recommended )
+
+All the computations or operations in a graph will be specified with the same accumulation data type. Whether the given accumulation mode can be utilized is decided by the backend capbility. 
+
+This option refers to the current solution for floating-point match mode attribute support in oneDNN Graph. These attributes share some similarities to a certain extent: both provide a certain degree of speedup by modifying the datatype based on backend capabilities.
+
+The API will be like:
+
+```cpp
+// for demonstrate purpose, we may provide the graph parameter in a wrapper or define more
+// constructors. please refer to the implementation part for a brief introduction.
+graph g(kind, accumulation_mode);
```

Comment:
OK. Will confirm with oneDNN about the behavior. Since oneDNN does not support to check whether the given `accumulation mode` is compatible currently, I think most likely users will receive an exception for unsupported cases due to creating primitive failure during compilation.

Reply:
@TaoLv Both serve different purposes:
- fpmath_mode allows conversion of inputs during computation
- accumulation_mode allows conversion of intermediate accumulators.

In the example you gave, a primitive implementation would be allowed to convert inputs to bf16 (in order to use AMX/XMX instructions), and also use f16 for accumulators (e.g. in instructions or when writing partial accumulators to memory)